File: ormdsl-js/ast.js
/*
 * Constrained set
 */
class Set {
constructor(values) {
this.values = values ? values : [];  // list of values
this.constraintSets = []; // one constraint set for each value
this.values.map(() => this.constraintSets.push([]));
this.constraints = [];    // constraint set for all values
}

// add constraint to each value
filter(f) {
for(var i=0; i<this.values.length; i++) {
this.constraintSets[i].push(f(this.values[i]));
}
return this;
}
// add constraint for all value
assert(f) {
this.constraints.push(f(this.values));
return this;
}
// return a decision variable
get(index) {
return new Var(this, index);
}
// make a copy -- internal use
clone() {
const set = new Set(this.values);
for(var i=0; i<this.constraintSets.length; i++) {
this.constraintSets[i].forEach (c => set.constraintSets[i].push(c));
}
this.constraints.forEach(c => set.constraints.push(c));
return set;
}
// make a new set using f
map(f) {
const set = this.clone();
set.values = set.values.map(v => f(v));
return set;
}
}

// fixed binary set
Set.binary = new Set([0,1]);

// dummy index
class Index {}

class Exp {
lt(that) { return new Comparison(this, '<=', that); }
gt(that) { return new Comparison(this, '>=', that); }
eq(that) { return new Comparison(this, '==', that); }

plus(that) { return new Plus(this, that); }
times(that) { return new Times(this, that); }
}

/*
 * Decision variable
 */
class Var extends Exp {
// set: constrained set
// index: either a number or undefined (so the index will an index object)
constructor(set, index) {
super();
this.set = set;
this.index = index ? index : new Index();
}
// make a new decision variable using f
map (f) {
return new Var(this.set.map(f), this.index);
}
}

class Plus extends Exp {
constructor(left, right) {
super();
this.left = left;
this.right = right;
}
}
class Times extends Exp {
constructor(left, right) {
super();
this.left = left;
this.right = right;
}
}

class Comparison {
constructor(left, op, right) {
this.left = left;
this.op = op;
this.right = right;
}
}

/*******  example application **************/


const N = 5; // number of warehouses
const M = 7 // number of consumers
const P = 4; // number of established warehouses

const distances = [];

var until = (from, to) => {
var ret = [];
for(var x=from; x<to; x++) ret.push(x);
return ret;
}

class Warehouse {
constructor(id) {
this.id = id;
this.established = Set.binary.get();
}
}

// weight: int
// distances: int[] -- distances from this consumer to each warehouse
// warehouse: decision variable of Warehouse type
class Consumer {
constructor(weight, distances, warehouse) {
this.weight = weight;
this.distances = distances;
this.warehouse = warehouse;
}

factor() {
return this.warehouse.map(w => this.distances[w.id]).times(this.weight);
}
}

const warehouses = until(0, N).map(j => new Warehouse(j))

until(0, M).forEach(i => {
distances[i] = [];
until(0, N).forEach(j => {
distances[i][warehouses[j].id] = Math.round(Math.random() * 10);
})
});

const weights = [];

until(0, M).forEach(i => { weights[i] = Math.round(Math.random() * 10); });

const warehouseSet = 
new Set(warehouses)
.assert( // make sure total established warehouses is less than P
ws => ws.map(w => w.established).reduce((a,b) => a.plus(b)).lt(P)
) // make sure that only established warehouses are considered
.filter(w => w.established.eq(1));

const consumers = until(0, M).map(i => new Consumer(weights[i], distances[i], warehouseSet.get()));

const objective = consumers.map(c => c.factor()).reduce((a,b) => a.plus(b));

File: ormdsl-py/ast.py
# -*-coding:utf-8-*-
from functools import reduce
import random
count = 0
def getName():
    global count
    count += 1
    return 'x' + str(count)


class Exp:
    def __init__(self):
        pass

    def lt(self, that):
        return Comparison(self,'<=', that)

    def gt(self, that):
        return Comparison(self, '>=', that)

    def eq(self, that):
        return Comparison(self, '==', that)

    def plus(self, that):
        return Plus(self, that)

    def times(self, that):
        return Times(self, that)


# binary decision variable
class Index(Exp):
    def __init__(self):
        self.name = getName()
    
    def toString(self):
         return self.name
   

# Decision variable
class Var(Exp):
    # set: constrained set
    # index: either a number or undefined 
    def __init__(self, Set, index, indicators):
        self.set = Set
        self.index = index
        if not index and not indicators:
            self.indicators = map(lambda x:Index(), self.set)
        elif not index and indicators:
            self.indicators = indicators
        
    
    def getConstraints(self):
        return self.set.getConstraints() + [] if self.index else map(lambda e, i: e.gt(self.set.indicators[i]), self.indicators)
    

    # make a decision variable using f
    def map(self, f):
        return Var(self.set.map(f), self.index, self.indicators)
    

    def toString(self):
        return '(' + ( self.set.indicators[self.index].times(self.set.values[self.index]) if self.index else reduce(lambda a, b: a.plus(b),map(lambda e, i:e.times(self.set.values[i]),self.indicators))) + ')'


class Plus(Exp):
    def __init__(self, left, right):
        self.left = left
        self.right = right
    

    def toString(self):
        return self.left.toString() + '+' + self.right.toString()
    

    def getConstraints(self):
        left = self.left.getConstraints() if isinstance(self.left, Exp) else []
        right = self.right.getConstraints() if isinstance(self.right, Exp) else []
        return left + right
    

class Times(Exp):
    def __init__(self, left, right):
        super()
        self.left = left
        self.right = right

    def toString(self):
        return self.left.toString() + '*' + self.right.toString()

    def getConstraints(self):
        left =  self.left.getConstraints() if  isinstance(self.left, Exp) else []
        right = self.right.getConstraints() if isinstance(self.right, Exp) else []
        return left + right


class Implication:
    def __init__(self,left, right):
        self.left = left
        self.right = right
    
    def toString(self):
        return self.left.toString() + ' => ' + self.right.toString()
    

class Comparison:
    def __init__(self,left, op, right):
        self.left = left
        self.op = op
        self.right = right


    def toString(self):
        return self.left.toString() + self.op + self.right.toString()


# Constrained set
class Set:
    def __init__(self, values):
        if not values:
            raise Exception("set cannot be empty")
        self.values = values  # list of values
        self.constraintSets = [] # one constraint set for each value
        map(lambda x: self.constraintSets.append([]), self.values)
        self.constraints = []    # constraint set for all values
        self.indicators = list(map(lambda x: Index(), self.values)) # a list of indicator variable names for constraint output 

    def getConstraints(self):
        ret = []
        map(lambda cs, i: map(lambda c: ret.append(Implication(Comparison(self.indicators[i], '==', 1), c)), cs), self.constraintSets)
        return ret + self.constraints


    def toString(self):
        return reduce(lambda a,b: a + '\n' + b, self.getConstraints())


    # add constraint to each value
    def filter(self, f):
        for i in range(len(self.values)):
            self.constraintSets.append([])
            self.constraintSets[i].append(f(self.values[i]))
        return self


    # add constraint for all value
    def Assert(self,f):
        self.constraints.append(f(self.values))
        return self


    # return a decision variable    
    def get(self, index):
        return Var(self, index, None)


    # make a copy -- internal use
    def clone(self):
        S = Set(self.values)
        map(lambda c: S.constraintSets.append(c), self.constraintSets)
        map(lambda c: S.constraints.append(c), self.constraints)
        return S


    # make a set using f
    def map(self, f):
        Set = self.clone()
        Set.values = map(lambda v:f(v), Set)
        return Set


#*******  example application **************


N = 4 # number of warehouses
M = 5 # number of consumers
P = 3 # number of established warehouses



class Warehouse:
    def __init__(self, d):
        self.id = id
        self.established = Index()


# weight: int
# distances: int[] -- distances from self consumer to each warehouse
# warehouse: decision variable of Warehouse type
class Consumer:
    def __init__(self, id, weight, distances, warehouse):
        self.id = id
        self.weight = weight
        self.distances = distances
        self.warehouse = warehouse

    def factor(self):
        return self.warehouse.map(lambda w: self.distances[w.id]).times(self.weight)

warehouses = [Warehouse(j) for j in range(N)]

distances=[]
for i in range(M):
    distances.append([])
    for j in range(N):
        distances[i].append(round(random.random() * 10))

weights = []
for i in range(M):
    weights.append(round(random.random() * 10))

# make sure total established warehouses is less than P
# make sure that only established warehouses are considered
warehouseSet = Set(warehouses).Assert(lambda ws: reduce(lambda a,b: a.plus(b), map(lambda w: w.established, ws)).lt("P")).filter(lambda w: w.established.eq(1))

consumers = [Consumer(i, weights[i], distances[i], warehouseSet.get()) for i in range(N)]

objective = reduce(lambda a, b: a.plus(b), consumers.map(lambda c: c.factor()))

print(objective.toString())

print(reduce(lambda a, b: a + '\n' + b, objective.getConstraints()))